posts/non-fungible-tokens-on-the-ethereum-blockchain.md

```diff
@@ -0,0 +1,523 @@
+    if (txReceipt) {
+        this.setState({
+            myToken: {
+                ...this.state.myToken,
+                txReceipt: txReceipt
+            }
+        });
+    }
+});
+```
+Note the `await` modifiers; Ethereum transactions are not real-time, so these are async calls.
+
+**Caveat #13,427:** you must have been added as a minter for this token using the `addMinter` function!
+
+Here's the code for adding a minter to our token (though not currently exposed on ALF):
+```javascript
+// grab user's metamask account address
+const accounts = await web3.eth.getAccounts();
+const myTokenInstance = new web3.eth.Contract(
+    JSON.parse(this.state.myToken.abi),
+    this.state.web3ctx.tokenAddress
+);
+
+console.log('Adding minter address ' + minterAddress + ' as minter for token address '
+    + myTokenInstance.options.address);
+
+// see, this https://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#methods-mymethod-send
+await myTokenInstance.methods.addMinter(minterAddress).send({
+    from: accounts[0]
+}, (error, txHash) => {
+    console.log('txHash = ' + txHash + '; Error = ' + error);
+});
+```
+
+Once we've minted our token, the status block provides a link to our transaction record on https://etherscan.io and shows the tx hash, block number, and gas used:
+
+![TxData](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-10.png)
+
+I waited a few minutes for my transaction to be mined on Ethereum, then I head to https://opensea.io/account and my shiny new ERC-721 token has appeared!
+
+![OpenSea](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-11.png)
+
+## Conclusion
+Non-fungible tokens present some interesting opportunities in tokenization; fungibility isn't for everything.
+
+By the same token...
+
+![BaDumTss](https://raw.githubusercontent.com/misterzero/blog-usa/master/images/2019/01/non-fungible-tokens-12.png)
+
+...not everything needs to be tokenized. It's fun to work with the pioneers from the art and gaming worlds and explore how the technology can be applied to both digital and physical assets.
+
+I find it a bit inconsistent that we're minting tokens on a distributed ledger using a distributed file system, yet we're using old school centralized DNS to reference our metadata and images via IPFS gateways.
+
+This could be remedied with an [oracle (not that Oracle!) on the blockchain](https://medium.com/@jesus_notchrist/blockchain-oracles-af3b216bed6b) to locate the metadata and image files. If this was implemented, then perhaps the organizations involved with NFTs would agree to a new standard where the **IPFS hash** of the metadata is stored in the token, rather than the URI, and similarly the IPFS hash of the image is stored in the metadata. This way, we could at least hedge our reliance on a single DNS domain by allowing the oracle to determine which IPFS gateway is appropriate at the time the image is requested.
```

Comment:
```suggestion
This could be remedied with an [oracle (not that Oracle!) on the blockchain](https://medium.com/@jesus_notchrist/blockchain-oracles-af3b216bed6b) to locate the metadata and image files. If this was implemented, then perhaps the organizations involved with NFTs would agree to a new standard where the **IPFS hash** of the metadata is stored in the token, rather than the URI, and similarly, the IPFS hash of the image is stored in the metadata. This way, we could at least hedge our reliance on a single DNS domain by allowing the oracle to determine which IPFS gateway is appropriate at the time the image is requested.
```